### ISSUE TITLE
Incorrect sort result after join

### ISSUE THREAD

**Issue by LiuYangkuan** [ClickHouse/ClickHouse#51660]
> You have to provide the following information whenever possible.

We have two tables with data below:
```
create table test1 (
    `pt` String,
    `brand_name` String,
    `total_indirect_order_cnt` Float64,
    `total_indirect_gmv` Float64
) ENGINE = Memory

create table test2 (
    `pt` String,
    `brand_name` String,
    `exposure_uv` Float64,
    `click_uv` Float64
) ENGINE = Memory

INSERT INTO test1 (`pt`, `brand_name`, `total_indirect_order_cnt`, `total_indirect_gmv`) VALUES
('20230625', 'LINING', 2232, 1008710),
('20230625', 'adidas', 125, 58820),
('20230625', 'Nike', 1291, 1033020), 
('20230626', 'Nike', 1145, 938926),
('20230626', 'LINING', 1904, 853336),
('20230626', 'adidas', 133, 62546),
('20220626', 'LINING', 3747, 1855203),
('20220626', 'Nike', 2295, 1742665), 
('20220626', 'adidas', 302, 122388);

INSERT INTO test2 (`pt`, `brand_name`, `exposure_uv`, `click_uv`) VALUES 
('20230625', 'Nike', 2012913, 612831),
 ('20230625', 'adidas', 480277, 96176),
 ('20230625', 'LINING', 2474234, 627814),
 ('20230626', 'Nike', 1934666, 610770),
 ('20230626', 'adidas', 469904, 91117),
 ('20230626', 'LINING', 2285142, 599765),
 ('20220626', 'Nike', 2979656, 937166),
 ('20220626', 'adidas', 704751, 124250),
 ('20220626', 'LINING', 3163884, 1010221);
```

When we query using sql below:
```
SELECT * FROM (
        SELECT  m0.pt                                                                                     AS pt
               ,m0.`uvctr`                                                                                AS uvctr
               ,round(m1.uvctr,4)                                                                         AS uvctr_hb_last_value
               ,round(m2.uvctr,4)                                                                         AS uvctr_tb_last_value
        FROM
        (
                SELECT  m0.pt                                                                                                           AS pt
                       ,COALESCE(m0.brand_name,m1.brand_name)                                                                           AS brand_name
                       ,if(isNaN(`click_uv` / `exposure_uv`) OR isInfinite(`click_uv` / `exposure_uv`),NULL,`click_uv` / `exposure_uv`) AS `uvctr`
                FROM
                (
                                SELECT  pt          AS pt
                                       ,brand_name  AS `brand_name`
                                       ,exposure_uv AS `exposure_uv`
                                       ,click_uv    AS `click_uv`
                                FROM test2
                                WHERE pt = '20230626'
                ) m0
                FULL JOIN
                (
                                SELECT  pt                        AS pt
                                       ,brand_name                AS `brand_name`
                                       ,total_indirect_order_cnt  AS `total_indirect_order_cnt`
                                       ,total_indirect_gmv        AS `total_indirect_gmv`
                                FROM test1
                                WHERE pt = '20230626'
                ) m1
                ON m0.brand_name = m1.brand_name AND m0.pt = m1.pt
        ) m0
        LEFT JOIN
        (
                SELECT  m0.pt AS pt
                       ,if(isNaN(`click_uv` / `exposure_uv`) OR isInfinite(`click_uv` / `exposure_uv`),NULL,`click_uv` / `exposure_uv`) AS `uvctr`
                       ,COALESCE(m0.brand_name,m1.brand_name)                                                                 AS brand_name
                       ,`exposure_uv`                                                                                         AS `exposure_uv`
                       ,`click_uv`
                FROM
                (
                                SELECT  pt          AS pt
                                       ,brand_name  AS `brand_name`
                                       ,exposure_uv AS `exposure_uv`
                                       ,click_uv    AS `click_uv`
                                FROM test2
                                WHERE pt = '20230625'
                ) m0
                FULL JOIN
                (
                                SELECT  pt                       AS pt
                                       ,brand_name               AS `brand_name`
                                       ,total_indirect_order_cnt AS `total_indirect_order_cnt`
                                       ,total_indirect_gmv       AS `total_indirect_gmv`
                                FROM test1
                                WHERE pt = '20230625'
                ) m1
                ON m0.brand_name = m1.brand_name AND m0.pt = m1.pt
        ) m1
        ON m0.brand_name = m1.brand_name AND m0.pt = m1.pt
        LEFT JOIN
        (
                SELECT  m0.pt AS pt
                       ,if(isNaN(`click_uv` / `exposure_uv`) OR isInfinite(`click_uv` / `exposure_uv`),NULL,`click_uv` / `exposure_uv`) AS `uvctr`
                       ,COALESCE(m0.brand_name,m1.brand_name)                                                                 AS brand_name
                       ,`exposure_uv`                                                                                         AS `exposure_uv`
                       ,`click_uv`
                FROM
                (
                                SELECT  pt          AS pt
                                       ,brand_name  AS `brand_name`
                                       ,exposure_uv AS `exposure_uv`
                                       ,click_uv    AS `click_uv`
                                FROM test2
                                WHERE pt = '20220626'
                ) m0
                FULL JOIN
                (
                                SELECT  pt                        AS pt
                                       ,brand_name                AS `brand_name`
                                       ,total_indirect_order_cnt  AS `total_indirect_order_cnt`
                                       ,total_indirect_gmv        AS `total_indirect_gmv`
                                FROM test1
                                WHERE pt = '20220626'
                ) m1
                ON m0.brand_name = m1.brand_name AND m0.pt = m1.pt
        ) m2
        ON m0.brand_name = m2.brand_name AND m0.pt = m2.pt
) c0
ORDER BY pt ASC, uvctr DESC
```
But the result is below:
```
20230626        0.3156979034107179      \N      \N
20230626        0.19390556368960468     \N      \N
20230626        0.2624629016490004      \N      \N
```

**Comment by qoega** [comment]
I think issue is related to some identifiers. It is not reproducing with analyzer

```
...
SETTINGS allow_experimental_analyzer = 1

Query id: dcefeec9-2e27-46f5-ab8e-6a9b48894c10

┌─pt───────┬───────────────uvctr─┬─uvctr_hb_last_value─┬─uvctr_tb_last_value─┐
│ 20230626 │  0.3156979034107179 │                ᴺᵁᴸᴸ │                ᴺᵁᴸᴸ │
│ 20230626 │  0.2624629016490004 │                ᴺᵁᴸᴸ │                ᴺᵁᴸᴸ │
│ 20230626 │ 0.19390556368960468 │                ᴺᵁᴸᴸ │                ᴺᵁᴸᴸ │
└──────────┴─────────────────────┴─────────────────────┴─────────────────────┘

3 rows in set. Elapsed: 0.015 sec.
```

So it will be fixed with #42648. You can enable analyzer for your queries even now if they work as expected. There is less than a 1% tests that are not yet fixed so most cases work already.

**Comment by vdimir** [comment]
Can be resolved with https://github.com/ClickHouse/ClickHouse/pull/51419

**Comment by qoega** [comment]
Also you can just change `AS uvctr` to `AS uvctr_` and updated order by accordingly.

**Comment by LiuYangkuan** [comment]
> Also you can just change `AS uvctr` to `AS uvctr_` and updated order by accordingly.

Our sql is generated by programs, we can't edit it by manual.

**Comment by LiuYangkuan** [comment]
> I think issue is related to some identifiers. It is not reproducing with analyzer
> 
> ```
> ...
> SETTINGS allow_experimental_analyzer = 1
> 
> Query id: dcefeec9-2e27-46f5-ab8e-6a9b48894c10
> 
> ┌─pt───────┬───────────────uvctr─┬─uvctr_hb_last_value─┬─uvctr_tb_last_value─┐
> │ 20230626 │  0.3156979034107179 │                ᴺᵁᴸᴸ │                ᴺᵁᴸᴸ │
> │ 20230626 │  0.2624629016490004 │                ᴺᵁᴸᴸ │                ᴺᵁᴸᴸ │
> │ 20230626 │ 0.19390556368960468 │                ᴺᵁᴸᴸ │                ᴺᵁᴸᴸ │
> └──────────┴─────────────────────┴─────────────────────┴─────────────────────┘
> 
> 3 rows in set. Elapsed: 0.015 sec.
> ```
> 
> So it will be fixed with #42648. You can enable analyzer for your queries even now if they work as expected. There is less than a 1% tests that are not yet fixed so most cases work already.

Our version is 22.8, haven't support settings allow_experimental_analyzer yet.
